applications/connectivity_bridge/src/modules/usb_bulk_commands.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/retention/bootmode.h>
+#include <zephyr/sys/reboot.h>
+#include <zephyr/drivers/gpio.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bulk_commands, CONFIG_BRIDGE_BULK_LOG_LEVEL);
+
+#define ID_DAP_VENDOR14		       (0x80 + 14)
+#define ID_DAP_VENDOR15		       (0x80 + 15)
+#define ID_DAP_VENDOR_NRF53_BOOTLOADER ID_DAP_VENDOR14
+#define ID_DAP_VENDOR_NRF91_BOOTLOADER ID_DAP_VENDOR15
+
+static const struct gpio_dt_spec reset_pin =
+	GPIO_DT_SPEC_GET_OR(DT_PATH(zephyr_user), reset_gpios, {});
+static const struct gpio_dt_spec button1_pin =
+	GPIO_DT_SPEC_GET_OR(DT_PATH(zephyr_user), button1_gpios, {});
+
+static void nrf91_bootloader_work_handler(struct k_work *work);
+K_WORK_DELAYABLE_DEFINE(nrf91_bootloader_work, nrf91_bootloader_work_handler);
+
+static void nrf91_bootloader_work_handler(struct k_work *work)
+{
+	ARG_UNUSED(work);
+	if (reset_pin.port && device_is_ready(reset_pin.port)) {
+		LOG_DBG("setting reset pin");
+		gpio_pin_configure_dt(&reset_pin, GPIO_OUTPUT_LOW);
+	}
+	if (button1_pin.port && device_is_ready(button1_pin.port)) {
+		LOG_DBG("setting button pin");
+		gpio_pin_configure_dt(&button1_pin, GPIO_OUTPUT_LOW);
+	}
+	k_sleep(K_MSEC(100));
+	if (reset_pin.port && device_is_ready(reset_pin.port)) {
+		gpio_pin_configure_dt(&reset_pin, GPIO_DISCONNECTED);
+	}
+	k_sleep(K_MSEC(1000));
+	if (button1_pin.port && device_is_ready(button1_pin.port)) {
+		gpio_pin_configure_dt(&button1_pin, GPIO_DISCONNECTED);
+	}
+}
+
+/* The is a placeholder implementation until proper CMSIS-DAP support is available. */
+size_t dap_execute_cmd(uint8_t *in, uint8_t *out)
+{
+	LOG_DBG("got command 0x%02X", in[0]);
+
+#ifdef CONFIG_RETENTION_BOOT_MODE
+	if (in[0] == ID_DAP_VENDOR_NRF53_BOOTLOADER) {
+		bootmode_set(BOOT_MODE_TYPE_BOOTLOADER);
+		sys_reboot(SYS_REBOOT_WARM);
```

Comment:
This is a highly critical feature. I don't like the idea of having it depend on complex code.

Reply:
> This shutdown feature was puzzling for me since it is seemingly never used and also does not make sense for the application. There is no point in shutting down this device without restarting it or at least defining an interrupt to trigger start-up.

This application should enter system off mode to save power when USB is disconnected (unless BLE is enabled). A voltage on VBUS will trigger a wakeup from system off. At least this was the behavior on Thingy:91 when i worked on this. I haven't tested on Thingy:91 X.

> This is a highly critical feature. I don't like the idea of having it depend on complex code.

The USB subsystem and this class is already complex code, so I don't understand why doing a clean reboot is such a comparatively big risk. Either way, I am mostly concerned about the reboot happening in the middle of a flash write. But any flash write via `config.c` should happen only after a USB disconnect, and from the system workqueue priority, so I think this sudden reboot cannot currently happen at the same time.

I'm not very familiar with Thingy:91 X, but have you considered if there are e.g. pin configurations, LEDs, or similar, that should be set before doing a firmware update? If there isn't, I'll be OK doing this way for now and switching to a more systematic way of handling reset if the need should arise in the future